MdeModulePkg/Core/Dxe/Misc/MemoryProtectionSupport.c

```diff
@@ -0,0 +1,2853 @@
+
+  return EFI_SUCCESS;
+}
+
+STATIC MEMORY_PROTECTION_SPECIAL_REGION_PROTOCOL  mMemoryProtectionSpecialRegion =
+{
+  GetSpecialRegions,
+  AddSpecialRegion
+};
+
+/**
+  Initialize the memory protection special region reporting.
+**/
+VOID
+EFIAPI
+CoreInitializeMemoryProtectionSpecialRegions (
+  VOID
+  )
+{
+  EFI_STATUS  Status;
+  EFI_HANDLE  HandleNull = NULL;
+
+  Status = CollectSpecialRegionHobs ();
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a - Failed to collect the memory protection special region HOB entries!\n",
+      __FUNCTION__
+      ));
+  }
+
+  Status = CoreInstallMultipleProtocolInterfaces (
+             &HandleNull,
+             &gMemoryProtectionSpecialRegionProtocolGuid,
+             &mMemoryProtectionSpecialRegion,
+             NULL
+             );
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a - Failed to install the memory protection special region protocol!\n",
+      __FUNCTION__
+      ));
+  }
+
+  return;
+}
+
+// ---------------------------------------
+//       USEFUL DEBUG FUNCTIONS
+// ---------------------------------------
+
+/**
+  Debug dumps the memory map.
+
+  @param[in]  MemoryMapSize     A pointer to the size, in bytes, of the MemoryMap buffer
+  @param[in]  MemoryMap         A pointer to the buffer containing the memory map
+  @param[in]  DescriptorSize    Size, in bytes, of an individual EFI_MEMORY_DESCRIPTOR
+**/
+STATIC
+VOID
+DumpMemoryMap (
+  IN CONST UINTN             *MemoryMapSize,
+  IN  EFI_MEMORY_DESCRIPTOR  *MemoryMap,
+  IN CONST UINTN             *DescriptorSize
+  )
+{
+  EFI_MEMORY_DESCRIPTOR  *MemoryMapEntry;
+  EFI_MEMORY_DESCRIPTOR  *MemoryMapEnd;
+
+  if ((MemoryMapSize == NULL) || (MemoryMap == NULL) || (DescriptorSize == NULL)) {
+    return;
+  }
+
+  MemoryMapEntry = MemoryMap;
+  MemoryMapEnd   = (EFI_MEMORY_DESCRIPTOR *)((UINT8 *)MemoryMap + *MemoryMapSize);
+  while (MemoryMapEntry < MemoryMapEnd) {
+    DEBUG ((
+      DEBUG_INFO,
+      "Memory Range: 0x%llx - 0x%llx. Type:%d, Attributes: 0x%llx\n",
+      MemoryMapEntry->PhysicalStart,
+      MemoryMapEntry->PhysicalStart + EfiPagesToSize (MemoryMapEntry->NumberOfPages),
+      MemoryMapEntry->Type,
+      MemoryMapEntry->Attribute
+      ));
+    MemoryMapEntry = NEXT_MEMORY_DESCRIPTOR (MemoryMapEntry, *DescriptorSize);
+  }
+}
+
+/**
+  Debug dumps the input bitmap
+
+  @param[in] Bitmap Pointer to the start of the bitmap
+  @param[in] Count  Number of bitmap entries
+**/
+STATIC
+VOID
+DumpBitmap (
+  IN CONST UINT8  *Bitmap,
+  IN UINTN        Count
+  )
+{
+  UINTN  Index = 0;
+
+  DEBUG ((DEBUG_INFO, "Bitmap: "));
+  for ( ; Index < Count; Index++) {
+    DEBUG ((DEBUG_INFO, "%d", IS_BITMAP_INDEX_SET (Bitmap, Index) ? 1 : 0));
+  }
+
+  DEBUG ((DEBUG_INFO, "\n"));
+}
+
+/**
+  Debug dumps the input bitmap
+
+  @param[in] Bitmap Pointer to the start of the bitmap
+  @param[in] Count  Number of bitmap entries
+**/

```

Comment:
This function description is the same as `DumpBitmap()` and needs to be updated.